extensions/guacamole-auth-nextcloud/src/main/java/org/apache/guacamole/auth/nextcloud/NextcloudJwtAuthenticationProvider.java

```diff
@@ -0,0 +1,189 @@
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.interfaces.ECPublicKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.Base64;
+import java.util.Date;
+import javax.servlet.http.HttpServletRequest;
+
+import org.apache.guacamole.GuacamoleException;
+import org.apache.guacamole.net.auth.AbstractAuthenticationProvider;
+import org.apache.guacamole.net.auth.AuthenticatedUser;
+import org.apache.guacamole.net.auth.Credentials;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Allows a pre-check of users with encrypted Nextcloud JWT data blocks.
+ * The username in the JWT will be compared with a list in guacamole.properties.
+ * The JWT will be verified with the public key. If the JWT is valid, the login
+ * page will be loaded. If the JWT is missing or invalid, an exception message
+ * will be displayed.
+ */
+public class NextcloudJwtAuthenticationProvider extends AbstractAuthenticationProvider {
+
+    private static final int MINUTES_TOKEN_VALID = 1;
+    /**
+     * Injector which will manage the object graph of this authentication
+     * provider.
+     */
+    private final Injector injector;
+
+    @Inject
+    private ConfigurationService confService;
+
+    private static final Logger logger = LoggerFactory.getLogger(NextcloudJwtAuthenticationProvider.class);
+
+    /**
+     * Creates a new MextcloudJwtAuthenticationProvider that authenticates user.
+     *
+     * @throws GuacamoleException
+     *     If a required property is missing, or an error occurs while parsing
+     *     a property.
+     */
+    public NextcloudJwtAuthenticationProvider() throws GuacamoleException {
+
+        // Set up Guice injector.
+        injector = Guice.createInjector(new NextcloudJwtAuthenticationProviderModule(this));
+
+    }
+
+    @Override
+    public String getIdentifier() {
+        return "nextcloud";
+    }
+
+    @Override
+    public AuthenticatedUser authenticateUser(Credentials credentials) throws GuacamoleException {
+
+        HttpServletRequest request = credentials.getRequest();
+
+        String token = request.getParameter("nctoken");
+        String ipaddr = request.getRemoteAddr();
+
+        boolean localAddr = this.validIpAddress(ipaddr);
+        if (localAddr) {
+            logger.info("Request from local address {}", ipaddr);
+            return null;
+        }
+
+        if (token == null) {
+            throw new GuacamoleException("Missing token.");
+        }
+
+        try {
+            boolean valid = this.isValidJWT(token);
+            if (!valid) {
+                throw new GuacamoleException("Token expired.");
+            }
+            logger.info("Token valid.");
+        } catch (final GuacamoleException ex) {
+            logger.error("Token validation failed.", ex);
+            throw new GuacamoleException(ex.getMessage());
+        }
+        return null;
```

Comment:
It might be nice to have some documentation throughout this function - simply comments that indicate what you're doing and why.